### PR TITLE
Code Cleanup

### PR DESCRIPTION
**Pull request by stieglma**
As promised in issue #169 here is the merge request with some code cleanup, involving fixing all kinds of rawtype warnings and removing some unused imports, there would be even more things to fix (unused variables ins tests, an unclosed resource in `ApproximateStringSearchDemo` and many missing `serialVersionUID` fields in serializable classes, but I am not sure if you intend this to be like that so I only fixed the first two mentioned things in this merge request.

### CODE REVIEW

File: core/src/main/java/smile/classification/RBFNetwork.java [hunk 1]
```diff
@@ -156,7 +156,7 @@ public Trainer(Metric<T> distance) {
          * @param rbf the radial basis function.
          * @param m the number of basis functions.
          */
-        public Trainer setRBF(RadialBasisFunction rbf, int m) {
+        public Trainer<T> setRBF(RadialBasisFunction rbf, int m) {
```
Comment (haifengl) [inline core/src/main/java/smile/classification/RBFNetwork.java]: What's the warning about this? My IDE never report warnings about it. And previously another community member scanned the whole code base with lint tools. He didn't report something like this. I am just wonder why and is it because of Java 8? Thanks!

**Comment by haifengl** [comment]
Thank you very much! I will merge the PR once I understand the warning.

File: core/src/main/java/smile/classification/RBFNetwork.java (hunk 1 above)
Comment (stieglma) [inline core/src/main/java/smile/classification/RBFNetwork.java]: At first, no this isn't a Java 8 feature, this is in place since Generics are, I think this is since 1.5.

What happens there? Well Trainer itself is a class with a generic type parameter T (c.f. line 127 in the code, later on used for the distance field (`Metric<T>`)) and as such it should not refer to the "raw type" `Trainer` (without generic parameter) as you loose compile-time safety about the types used there.
Think of e.g. Lists, when you have a list of strings you write `List<String>` and get only offered methods to add strings to this list, and of course you can also only retrieve strings from that List (no ints, doubles, other types). But an even better example is the method `subList` which returns a List of Strings again, if they did omit the generic type parameter here, only a list would be returned where you do not know of which type the elements are, you would get only Objects instead everywhere. (This is actually what happens at runtime due to type erasure, but this is another topic ;) )

So in your specific case, the reason is that you wouldn't know which type is inside the trainer instance afterwards besides by using casts, which are not necessary if generics are used consistently.

Comment (stieglma) [inline core/src/main/java/smile/classification/RBFNetwork.java]: One further notice, in eclipse many of these warnings are generated directly when you write the code due to the incremental compiler, I am not sure how IntelliJ IDEA is working there but I have in mind that it doesn't compile instantly when you write code but only on certain actions. What I am quite sure of, is that IntelliJ does not by default run the so callled "code inspection" all the time, you have to trigger it manually, and it has a vast amount of configuration options. They are quite powerful and should definitely find and report such things.

Comment (stieglma) [inline core/src/main/java/smile/classification/RBFNetwork.java]: Sounds like a reasonable default, but I never heard of that, in fact you could also write something like this `public <X> Trainer<X> setRBF(...)` which could potentially make sense if you created a new trainer instance of different type there, so I think even if omitting the generic type parameter is valid you should add it such that this does not lead to problems in future development.

I also just tried that:

```
public class Test<T> {
        T tmp;

        public Test setTmp(final T x) {
            tmp = x;
            return this;
        }
        
        public T getTmp() {
            return tmp;
        }

        public Test<T> setTmp2(T y) {
            tmp = y;
            return this;
        }
    }
```

Now it is perfectly fine to write something like `String str = new Test<String>().setTmp2("abc").getTmp();`
but writing `String str = new Test<String>().setTmp("abc").getTmp();` does only work when you add a cast to `String`

**Comment by haifengl** [comment]
Thanks!
